src/realm/query_expression.hpp

```diff
@@ -487,80 +488,156 @@ class ValueBase {
+            right_size = std::unique(right.begin(), right.end()) - right.begin();
+        }
+
+        if constexpr (realm::is_any_v<TCond, BeginsWith, BeginsWithIns, EndsWith, EndsWithIns, Contains, ContainsIns,
+                                      Like, LikeIns>) {
+            // The string operators have the arguments reversed so we have to iterate right in the
+            // outer loop as this is actually the left argument
+            auto left_matches = [&](const QueryValue& right_val) {
+                for (size_t i = 0; i < left_size; i++) {
+                    if (c(left[i], right_val)) {
+                        // match
+                        if (compare_left == Compare::Any) {
+                            return true;
+                        }
+                        if (compare_left == Compare::None) {
+                            return false; // one matched
+                        }
+                    }
+                    else {
+                        // no match
+                        if (compare_left == Compare::All) {
+                            return false;
+                        }
+                    }
+                }
+                if (compare_left == Compare::None || compare_left == Compare::All) {
+                    return true;
+                }
+                return false;
+            };
+
+            for (size_t i = 0; i < right_size; i++) {
+                if (left_matches(right[i])) {
+                    if (compare_right == Compare::Any) {
                         return 0;
                     }
-                    if (right_cmp_type == ExpressionComparisonType::None) {
+                    if (compare_right == Compare::None) {
                         return not_found; // one matched
                     }
                 }
                 else {
-                    if (right_cmp_type == ExpressionComparisonType::All) {
+                    if (compare_right == Compare::All) {
                         return not_found;
                     }
                 }
             }
-            if (right_cmp_type == ExpressionComparisonType::None || right_cmp_type == ExpressionComparisonType::All) {
+            if (compare_right == Compare::None || compare_right == Compare::All) {
                 return 0; // either none or all
             }
         }
-        else if (left.m_from_link_list && !right.m_from_link_list) {
-            // Same as above, but with left values coming from link list.
-            REALM_ASSERT_DEBUG(right.size() > 0);
-            const size_t num_left_values = left.size();
-            ValueType right_val = right[0];
-            for (size_t l = 0; l < num_left_values; l++) {
-                bool match = c(left[l], right_val);
-                if (match) {
-                    if (left_cmp_type == ExpressionComparisonType::Any) {
+        else {
+            auto right_matches = [&](const QueryValue& left_val) {
```

Comment:
I agree that there is room for tuning, but I could not come up with an algorithm that would be optimal for all combinations of ALL, NONE and ANY. It also depends on if we should optimize for failing fast or for succeeding fast. Also, we are not always sorting - for small lists I think brute force is just as quick.

Reply:
To be more precise: I am convinced that in case of <, <=, >, >= then the whole thing can be boiled down to comparing just 2 values. First or last on left with first or last on right. But which ones to compare and if a true returned from the compare function should be seen as a match or a no match depends on the combination of ANY, ALL and NONE. On the other hand .., there is only 9 combinations.

Reply:
On the other hand I am not sure that the amortized complexity will be O(n*m) - in many cases we will have an early out. The current implementation has the same performance as before if one side is a single value. If we later find out which cases of list vs list it would make sense to optimize for, we can do that.

Reply:
Optimized the Equal and relational cases a bit.